### PR TITLE
Caching app information

### LINKED ISSUE #148: Add caching to the festival information, emergency info, and concert info

### PR DESCRIPTION
**Pull request by karimassi**
Closes #148

### CODE REVIEW

File: app/src/main/java/ch/epfl/balelecbud/BalelecbudApplication.java [hunk 1]
```diff
@@ -5,12 +5,17 @@
 
 import androidx.annotation.VisibleForTesting;
 
+import java.io.File;
```
Comment (Pialex99) [inline app/src/main/java/ch/epfl/balelecbud/BalelecbudApplication.java:8]: Doesn't seem to be used

File: app/src/main/java/ch/epfl/balelecbud/util/cache/FilesystemCache.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import ch.epfl.balelecbud.BalelecbudApplication;
+import ch.epfl.balelecbud.util.database.MyQuery;
+
+public class FilesystemCache implements Cache {
+
+    Gson gson;
+    Context context;
```
Comment (Pialex99) [inline app/src/main/java/ch/epfl/balelecbud/util/cache/FilesystemCache.java:28]:
```suggestion
    private Gson gson;
    private Context context;
```

File: app/src/androidTest/java/ch/epfl/balelecbud/util/database/CachedDatabaseTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import static ch.epfl.balelecbud.util.database.MockDatabase.assertQueryResults;
+import static ch.epfl.balelecbud.util.database.MockDatabase.karim;
+import static ch.epfl.balelecbud.util.database.MyWhereClause.Operator.EQUAL;
+
+@RunWith(AndroidJUnit4.class)
+public class CachedDatabaseTest {
+
+    private MockDatabase remoteDatabase;
+    private CachedDatabase appDatabase;
+    Cache appCache;
```
Comment (Pialex99) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/database/CachedDatabaseTest.java:43]:
```suggestion
    private Cache appCache;
```

File: app/src/androidTest/java/ch/epfl/balelecbud/util/database/CachedDatabaseTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    private FestivalInformation freshInfo = new FestivalInformation("Information", "Fresh information");
+    private FestivalInformation staleInfo = new FestivalInformation("Information", "Stale information");
+    private Map<String, Object> freshFriend = new HashMap<>();
+    private Map<String, Object> staleFriend = new HashMap<>();
```
Comment (Pialex99) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/database/CachedDatabaseTest.java:48]:
```suggestion
    private final FestivalInformation freshInfo = new FestivalInformation("Information", "Fresh information");
    private final FestivalInformation staleInfo = new FestivalInformation("Information", "Stale information");
    private final Map<String, Object> freshFriend = new HashMap<>();
    private final Map<String, Object> staleFriend = new HashMap<>();
```

File: app/src/main/java/ch/epfl/balelecbud/BalelecbudApplication.java [hunk 5]
```diff
@@ -20,6 +25,8 @@
 
     private static Context appContext;
     private static Database appDatabase;
+    private static Cache appCache;
```
Comment (Pialex99) [inline app/src/main/java/ch/epfl/balelecbud/BalelecbudApplication.java:28]: Does it really needs to be accessible from anywhere in the app ?

File: app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+//                sync.call();
+//            } else {
+//                sync.fail();
+//            }
+//        });
+//        sync.waitCall(1);
+//        sync.assertCalled(1);
+//        sync.assertNoFailedTests();
+//    }
+
```
Comment (Pialex99) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java:158]: Can be removed if not useful

File: app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import static ch.epfl.balelecbud.util.database.MockDatabase.karim;
+import static ch.epfl.balelecbud.util.database.MockDatabase.slot1;
+import static ch.epfl.balelecbud.util.database.MockDatabase.slot2;
+import static ch.epfl.balelecbud.util.database.MockDatabase.token1;
+import static ch.epfl.balelecbud.util.database.MyWhereClause.Operator.EQUAL;
+
+@RunWith(AndroidJUnit4.class)
+public class FilesystemCacheTest {
+
+    private Cache appCache = BalelecbudApplication.getAppCache();
```
Comment (Pialex99) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java:44]:
```suggestion
    private final Cache appCache = BalelecbudApplication.getAppCache();
```

File: app/src/main/java/ch/epfl/balelecbud/util/database/CachedDatabase.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        CompletableFuture<List<T>> result = CompletableFuture.completedFuture(new ArrayList<>());
+        if (cache.contains(query)) {
+            try {
+                result = cache.get(query, tClass);
+            } catch (FileNotFoundException e) {
+                Log.d(TAG, "queryWithType: " + e.getLocalizedMessage());
+            }
+        } else {
+            result = database.queryWithType(query, tClass);
+            result.whenComplete((ts, throwable) -> {
```
[the remark this answers is not in the document]
Comment (celinecamacho) [reply app/src/main/java/ch/epfl/balelecbud/util/database/CachedDatabase.java:55]: I agree, consider refactoring

File: app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        appCache.flush();
+        Assert.assertFalse(appCache.contains(new MyQuery(Database.USERS_PATH, Collections.emptyList())));
+        Assert.assertFalse(appCache.contains(new MyQuery(Database.LOCATIONS_PATH, Collections.emptyList())));
+        Assert.assertFalse(appCache.contains(new MyQuery(Database.CONCERT_SLOTS_PATH, Collections.emptyList())));
+        Assert.assertFalse(appCache.contains(new MyQuery(Database.TOKENS_PATH, Collections.emptyList())));
+
+    }
+
+
+//    private void assertMapResults(List<Map<String, Object>> expected,
```
Comment (celinecamacho) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java:141]: Delete these comments

File: app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    private Cache appCache = BalelecbudApplication.getAppCache();
+
+    @Before
+    public void setup() {
+        BalelecbudApplication.setRemoteDatabase(MockDatabase.getInstance());
+        appCache.flush();
+    }
+
+    @Test
+    public void containsCorrectForCollectionRetrieval() throws IOException {
```
Comment (celinecamacho) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java:53]: Avoid using "Correct" in test method names, you can rename to testContainsForCollectionRetrieval. Same for line 63

File: app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void containsWithFieldWhereClauseFalse() throws IOException {
+        appCache.put(Database.USERS_PATH, karim.getUid(), karim);
+        MyQuery query = new MyQuery(Database.USERS_PATH, new MyWhereClause("email", EQUAL, karim.getEmail()));
+        appCache.contains(query);
+    }
+
+    @Test
+    public void putGetWorksCorrectlyWithCustomTypes() throws Throwable {
```
Comment (celinecamacho) [inline app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java:86]: Also avoid using "WorksCorrectly" in test method names, you can rename to "testPutAndGetWithCustomTypes". Same for line 102

File: app/src/main/java/ch/epfl/balelecbud/util/database/CachedDatabase.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (karimassi) [reply app/src/main/java/ch/epfl/balelecbud/util/database/CachedDatabase.java:55]: While they're similar, it is not possible to refactor, as the types are not the same, and can't be rendered generic.

**Pull request merged**